do not compare cert bytes directly

irisdingbj opened the pull request:
We should not compare the root certs bytes directly, instead we can rely on the https://github.com/istio/istio/compare/master...irisdingbj:verifyfix?expand=1#diff-df16ac4e2cc1c048a4ce70aa3f27ec9c173573fc5288fb029fdc19b93c0ae259L133 to make sure the cert chain retured is correctly signed from root cert.

istio-testing commented:
@irisdingbj: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | e8acde757cbc777efdd1d8f26d9feb33c3c0b88f | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/37319/release-notes_istio/1493054859409100800) | true | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

irisdingbj commented:
/test release-notes_istio

istio-testing commented:
@irisdingbj: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | e8acde757cbc777efdd1d8f26d9feb33c3c0b88f | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/37319/release-notes_istio/1493054859409100800) | true | `/test release-notes_istio`
lint_istio | e8acde757cbc777efdd1d8f26d9feb33c3c0b88f | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/37319/lint_istio/1493054857681047552) | true | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@irisdingbj: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-pilot-multicluster_istio | 818b29f587e4ecfba50429654c92f4474baf3ee2 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/37319/integ-pilot-multicluster_istio/1493056308906037248) | true | `/test integ-pilot-multicluster_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

irisdingbj commented:
/test integ-pilot-multicluster_istio

istio-testing commented:
@irisdingbj: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-pilot-multicluster_istio | 818b29f587e4ecfba50429654c92f4474baf3ee2 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/37319/integ-pilot-multicluster_istio/1493079511535718400) | true | `/test integ-pilot-multicluster_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@irisdingbj: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-pilot-multicluster_istio | 818b29f587e4ecfba50429654c92f4474baf3ee2 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/37319/integ-pilot-multicluster_istio/1493079511535718400) | true | `/test integ-pilot-multicluster_istio`
lint_istio | 3f82da41aa45bedbd269ec3952f273484f7cd87a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/37319/lint_istio/1493087784615809024) | true | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@irisdingbj: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-pilot-multicluster_istio | 818b29f587e4ecfba50429654c92f4474baf3ee2 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/37319/integ-pilot-multicluster_istio/1493079511535718400) | true | `/test integ-pilot-multicluster_istio`
lint_istio | 3f82da41aa45bedbd269ec3952f273484f7cd87a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/37319/lint_istio/1493087784615809024) | true | `/test lint_istio`
gencheck_istio | 3f82da41aa45bedbd269ec3952f273484f7cd87a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/37319/gencheck_istio/1493087785437892608) | true | `/test gencheck_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn commented on security/pkg/pki/ra/k8s_ra.go:
```diff
@@ -116,24 +116,17 @@ func (r *KubernetesRA) SignWithCertChain(csrPEM []byte, certOpts ca.CertOpts) ([
-		}
-
-		if rootCertFromCertChain != nil {
+		} else if rootCertFromCertChain != nil {
```
this changes the behavior. before cert chain > mesh config, now its the inverse?

irisdingbj replied:
We now will verify cert-chain against the root cert in meshconfig not simply comare bytes. - in some cases, the root cert in cert-chain might be only an intermediate ca.

howardjohn reviewed (commented):
I don't think I really understand this, maybe @shankgan does

shankgan commented on security/pkg/pki/ra/k8s_ra.go:
```diff
@@ -116,24 +116,17 @@ func (r *KubernetesRA) SignWithCertChain(csrPEM []byte, certOpts ca.CertOpts) ([
 		if err != nil {
 			return nil, fmt.Errorf("failed to find root cert from mesh config (%v)", err.Error())
 		}
-		if rootCertFromCertChain != nil && rootCertFromMeshConfig != nil {
```
Some comments here would be super helpful. IMO, this logic is quite hard to read. Specifically, 

- IF there is a caCert specified in meshConfig, why we are we expecting that the CertChain of the signed cert from the external CA is signed by it
- What are the alternatives here - it appears that either the rootCert can be present in the certChain or not. What are the relevant scenarios where these could happen. Why does the MeshConfig root take precedence over the root in the certChain. 
- If we are extracting the "root" from the CertChain bytes, how come we are appending another "root" to it from MeshConfig. 
-

shankgan replied:
the logic does look correct to me though, still wondering if more comments + refactoring would make this more readable

shankgan commented on security/pkg/pki/ra/k8s_ra_test.go:
```diff
@@ -161,6 +161,41 @@ func TestK8sSignWithMeshConfig(t *testing.T) {
+	if err != nil {
+		t.Errorf("K8s CA Signing CSR With Root Cert In Meshconfig failed")
+	}
+}
```
Would be good to add a few more test cases here - 
(a) Scenario where certificate received from external CA is signed by a different root compared to meshConfig root and therefore certificate verification fails
(b)  Scenario where certificate received from external CA is signed by an intermediate cert that chains to the common root in the Meshconfig. IMO, this is what the newer code changes are for correct?

irisdingbj replied:
added more test cases

irisdingbj commented:
@howardjohn @shankgan  Thanks a lot for the review comments.  I have added comments to explain the possible scenarios in the code and also added more test cases to cover these secnarios.    The new code are majorly to handle the situation that cert chain contains only intermediate CA. So we can not compare bytes directly. PTAL and let me know if the current rules are good.

istio-testing commented:
@irisdingbj: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests_istio | c4a269b01217d9fffb7d2118db67a0498eaba6c4 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/37319/unit-tests_istio/1495938274353483776) | true | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

irisdingbj commented:
/test unit-tests_istio

shankgan commented:
@irisdingbj - looks good with the additional comments and the test cases
